### PR TITLE
Nullables

### PR DESCRIPTION
**Pull request by Happypig375**

### CODE REVIEW

**Comment by WhiteBlackGoose** [comment]
Before you complete it: why do you need them?

**Comment by Happypig375** [comment]
```cs
#nullable disable
        var tensor = GenTensor<object>.CreateVector(null, null, null); // Happily pass nulls into creation
        tensor.ToString(); // BOOM!


#nullable enable
        tensor = GenTensor<object>.CreateVector(null, null, null); // Warn: null cannot be cast to object
        tensor.ToString();

#nullable enable
        tensor = GenTensor<object?>.CreateVector(null, null, null); // Also warn: Cannot pass nullable type into tensor
        tensor.ToString();

**Comment by Happypig375** [comment]
Alternatively, you can remove the `notnull` constraint and fix the implementation of ToString(), and Nullable Reference Types will continue to warn all places where a possible null dereference will occur.

**Comment by WhiteBlackGoose** [comment]
Why can't I pass nulls into tensor? Let's leave it up to the user.

**Comment by Happypig375** [comment]
```cs
        var tensor = GenTensor<object>.CreateVector(null, null, null); // Happily pass nulls into creation
        tensor.ToString(); // BOOM!

**Comment by WhiteBlackGoose** [comment]
ToString of a tensor doesn't (or shouldn't?) call yourType.ToString. Instead, it class ConstantsAndFunctions<YourType>.ToString(yourType). Mb I forgot to implement it :). But you get the point

**Comment by Happypig375** [comment]
Then still, enabling Nullable Reference Types reminds of accidental null calls like the one mentioned above.

**Comment by WhiteBlackGoose** [comment]
But the user might need null as a type, for example, if they want to leave it as a default, and then initialize later. And their ToString will return whatever they wrote there, intentionally. 

Btw you could even disable runtime exceptions in this lib

**Comment by Happypig375** [comment]
>But the user might need null as a type, for example, if they want to leave it as a default, and then initialize later. And their ToString will return whatever they wrote there, intentionally.

Nullable Reference Types still enable this. 
```cs
// Once the notnull constraint is removed
#nullable enable
        tensor = GenTensor<object?>.CreateVector(null, null, null); // OK
        tensor.ToString(); // OK once implementation is fixed
```
It's about the implementation of `ToString()` and other functions not calling methods on elements directly, or other code where Null Reference Exception can occur.
>Btw you could even disable runtime exceptions in this lib

But you cannot disable bugs. Exceptions are fast ways to track down bugs and without them debugging will become a lot more difficult.

**Comment by WhiteBlackGoose** [comment]
Potentially I never call object's methods. I'll fix ToString

> Nullable Reference Types still enable this.

Users will have to use nullables types as well to avoid warnings?

**Comment by Happypig375** [comment]
No. If users have legacy code that cannot adopt Nullable Reference Types, then it will be like C# 7. If they do, then they can see the nullable annotations on APIs can get null tracking.

File: .gitignore [hunk 1]
```diff
@@ -1,24 +1,362 @@
...
+healthchecksdb
+
+# Backup folder for Package Reference Convert tool in Visual Studio 2017
+MigrationBackup/
+
+# Ionide (cross platform F# VS Code tools) working folder
+.ionide/
+
+# Fody - auto-generated XML schema
+FodyWeavers.xsd
```
Comment (WhiteBlackGoose) [inline .gitignore:362]: Why some 300 extra lines for the same thing? I don't need them to be honest

File: GenericTensor/Core/Exceptions.cs [hunk 2]
```diff
@@ -38,7 +38,7 @@ public class InvalidShapeException : ArgumentException
         public InvalidShapeException(string msg) : base(msg) {}
         public InvalidShapeException() : base() {}
 
-        internal static void NeedTensorSquareMatrix<T>(GenTensor<T> m)
+        internal static void NeedTensorSquareMatrix<T>(GenTensor<T> m) where T : notnull
```
Comment (WhiteBlackGoose) [inline GenericTensor/Core/Exceptions.cs:41]: It can be null

File: GenericTensor/Core/Tensor/Tensor.cs [hunk 3]
```diff
@@ -83,25 +88,15 @@ private void Init(TensorShape dimensions)
...
-        /// </summary>
-        public GenTensor(TensorShape dimensions)
-            => Init(dimensions);
-
         /// <summary>
         /// Creates an empty tensor where each element is just created wrapper
         /// </summary>
-        public GenTensor(params int[] dimensions)
-            => Init(new TensorShape(dimensions));
+        public GenTensor(params int[] dimensions) : this(new TensorShape(dimensions)) { }
```
Comment (WhiteBlackGoose) [inline GenericTensor/Core/Tensor/Tensor.cs:100]: Oof that's how I do it, thanks :)

File: GenericTensor/Core/Tensor/Iterate.cs [hunk 4]
```diff
@@ -48,11 +48,12 @@ private void NextIndex(int[] indecies, int id)
         /// <summary>
         /// Iterate over array of indecies and a value in TPrimitive
         /// </summary>
-        public IEnumerable<(int[] index, T value)> Iterate()
+        public IEnumerator<(int[] Index, T Value)> GetEnumerator()
         {
             foreach (var ind in IterateOver(0))
                 yield return (ind, this.GetValueNoCheck(ind));
         }
+        System.Collections.IEnumerator System.Collections.IEnumerable.GetEnumerator() => GetEnumerator();
```
Comment (WhiteBlackGoose) [inline GenericTensor/Core/Tensor/Iterate.cs:56]: There should be no implicit iterators, everything should be as explicit as possible for this lib

File: GenericTensor/Functions/SquareMatrixFactory.cs [hunk 5]
```diff
@@ -4,17 +4,19 @@
 
 namespace GenericTensor.Functions
 {
-    internal static class SquareMatrixFactory<T>
+    internal static class SquareMatrixFactory<T> where T : notnull
```
Comment (WhiteBlackGoose) [inline GenericTensor/Functions/SquareMatrixFactory.cs:7]: Again, T can be nullable

File: GenericTensor/Functions/SquareMatrixFactory.cs [hunk 6]
```diff
@@ -4,17 +4,19 @@
...
         internal static GenTensor<T> GetMatrix(int diagLength)
         {
-            if (TensorTempFactorySquareMatrices is null)
-                TensorTempFactorySquareMatrices = new List<GenTensor<T>> {null};
-            for (int i = TensorTempFactorySquareMatrices.Count; i <= diagLength; i++)
-                TensorTempFactorySquareMatrices.Add(new GenTensor<T>(i, i));
-            return TensorTempFactorySquareMatrices[diagLength - 1];
+            if (diagLength <= tensorTempFactorySquareMatrices.Count + 2)
+                lock (tensorTempFactorySquareMatrices)
+                    if (diagLength <= tensorTempFactorySquareMatrices.Count + 2)
```
Comment (WhiteBlackGoose) [inline GenericTensor/Functions/SquareMatrixFactory.cs:16]: Those two lines are just the same? The one before and the one after lock().

File: .gitignore (hunk 1 above)
Comment (Happypig375) [reply .gitignore:362]: This is the standard .gitignore for Visual Studio. It automatically ignores any common auto-generated files. You will not have to keep track of additional projects and tools then update this file anymore.

File: GenericTensor/Core/Tensor/Iterate.cs (hunk 4 above)
Comment (Happypig375) [reply GenericTensor/Core/Tensor/Iterate.cs:56]: Non-idiomatic but OK

File: GenericTensor/Functions/SquareMatrixFactory.cs (hunk 6 above)
Comment (Happypig375) [reply GenericTensor/Functions/SquareMatrixFactory.cs:16]: https://en.wikipedia.org/wiki/Double-checked_locking#Usage_in_C#

**Comment by WhiteBlackGoose** [comment]
Thanks for contributing

**Pull request merged**